linalg/constraints.hpp

```diff
@@ -0,0 +1,404 @@
+// Copyright (c) 2010-2021, Lawrence Livermore National Security, LLC. Produced
+// at the Lawrence Livermore National Laboratory. All Rights reserved. See files
+// LICENSE and NOTICE for details. LLNL-CODE-806117.
+//
+// This file is part of the MFEM library. For more information and source code
+// availability visit https://mfem.org.
+//
+// MFEM is free software; you can redistribute it and/or modify it under the
+// terms of the BSD-3 license. We welcome feedback and contributions, see file
+// CONTRIBUTING.md for details.
+
+#ifndef MFEM_CONSTRAINED
+#define MFEM_CONSTRAINED
+
+#include "solvers.hpp"
+#include "blockoperator.hpp"
+#include "sparsemat.hpp"
+#include "hypre.hpp"
+
+namespace mfem
+{
+
+class ParFiniteElementSpace;
+
+/** @brief An abstract class to solve the constrained system \f$ Ax = f \f$
+    subject to the constraint \f$ B x = r \f$.
+
+    Although implementations may not use the below formulation, for
+    understanding some of its methods and notation you can think of
+    it as solving the saddle-point system
+
+     (  A   B^T  )  ( x )          (  f  )
+     (  B        )  ( lambda )  =  (  r  )
+
+    Not to be confused with ConstrainedOperator, which is totally
+    different.
+
+    The height and width of this object as an IterativeSolver are for
```

Comment:
As many times people don't care about the lagrange multipliers, I would have the default `Mult` and on the primal variables and a possible extra `fullMult` or something like that for the multipliers.

Reply:
I first implemented it that way, but switched to this because it was used (in Smith) as the linear solver in a `NewtonSolver` object, which will always call with the `Mult()` method and has the saddle-point structure. I think that is probably the most expected use case?

Reply:
Hmmm. I'll think about this. When I wear my program hat, I think defaulting to only primal variables is more natural. When I wear my smith hat, I can see it the other way as you pointed out. Any thoughts @bslazarov @rcarson3 ?

Reply:
So from an application developer point of view, I personally would probably only care about the primal variables in order to advance my simulations. 

Although, I can't say I normally think of setting up my problems in this saddle-point kinda of way, so I can't say how much use I would have for the lagrange multipliers after the fact.